src/collider/src/collider.cpp

```diff
@@ -310,12 +309,17 @@ void Collider::UpdateBoxes(const VecDH<Box>& leafBB) {
   auto policy = autoPolicy(NumInternal());
   copy(policy, leafBB.cbegin(), leafBB.cend(), leaves.begin());
   // create global counters
-  VecDH<int> counter(NumInternal(), 0);
+  VecDH<int> states(NumLeaves(), 0);
+  sequence(policy, states.begin(), states.end());
+  for_each(policy, states.begin(), states.end(),
+           [] __host__ __device__(int& v) { v *= 2; });
+  int count = std::ceil(std::log2(nodeBBox_.size()));
```

Comment:
Hmm, you might want to have a look at the paper this is based on; it's a radix tree, not a balanced binary tree. It's been awhile, but if I recall the depth is limited to 32 (or maybe 64?). I'm too tired to give this a proper look tonight; I want to make sure I understand this; both my wrong assumption and what you're trying.

Reply:
Good catch, it is indeed not a balanced binary tree and its depth is at most 64. Actually I found that my explanation is wrong (I misunderstood the counter usage last night), will update it.